MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c

```diff
@@ -48,6 +55,233 @@ StatusToHexString (
+
+  if ((Certificate == NULL) || (IsFound == NULL)) {
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  *IsFound = FALSE;
+
+  //
+  // Read the size and attributes of the DB
+  //
+  Status = SystemTable->RuntimeServices->GetVariable (
+                                           L"db",
+                                           &gEfiImageSecurityDatabaseGuid,
+                                           &Attributes,
+                                           &DbSize,
+                                           NULL
+                                           );
+
+  //
+  // Confirm we got the expected error that the buffer was too small
+  // We need to know the size of the DB to allocate a buffer to read it
+  //
+  if (Status != EFI_BUFFER_TOO_SMALL) {
+    //
+    // Likely this will continue to fail on reboot
+    //
+    if (Status == EFI_SUCCESS) {
+      //
+      // While unlikely, it is possible that the DB is empty
+      // In that case, we should fail out gracefully because there is nothing we can do
+      //
+      Status = EFI_UNSUPPORTED;
+    }
+
+    goto Exit;
+  }
+
+  //
+  // Check that the DB has the expected attributes
+  //
+  if ((Attributes & VARIABLE_ATTRIBUTE_NV_BS_RT_AT) != VARIABLE_ATTRIBUTE_NV_BS_RT_AT) {
+    Status = EFI_UNSUPPORTED;
+    //
+    // Likely this will continue to fail on reboot
+    //
+    goto Exit;
+  }
+
+  Status = SystemTable->BootServices->AllocatePool (
+                                        EfiBootServicesData,
+                                        DbSize,
+                                        &Db
+                                        );
+  if ((Db == NULL) || EFI_ERROR (Status)) {
+    //
+    // Likely this will continue to fail on reboot
+    //
+    goto Exit;
+  }
+
+  //
+  // Grab the DB
+  //
+  Status = SystemTable->RuntimeServices->GetVariable (
+                                           L"db",
+                                           &gEfiImageSecurityDatabaseGuid,
+                                           &Attributes,
+                                           &DbSize,
+                                           Db
+                                           );
+  if (EFI_ERROR (Status)) {
+    //
+    // Likely this will continue to fail on reboot
+    // In that case, we should fail out gracefully because there is nothing we can do
+    //
+    goto Exit;
+  }
+
+  //
+  // Confirm the DB is at least as large as the EFI_SIGNATURE_LIST header
+  //
+  if (DbSize < sizeof (EFI_SIGNATURE_LIST)) {
+    Status = EFI_BUFFER_TOO_SMALL;
+    goto Exit;
+  }
+
+  //
+  // The DB is a list of EFI_SIGNATURE_LISTs but we only care about certificates
+  // they are not gauranteed to be in any particular order
+  //
+  CertList = (EFI_SIGNATURE_LIST *)Db;
+
+  //
+  // Iterate through the DB entries
+  // The DB is a list of EFI_SIGNATURE_LISTs
+  //
+  while ((DbSize > 0) && (DbSize >= CertList->SignatureListSize)) {

```

Comment:
Because `DbSize` is an unsigned integer type (`UINTN`), it is always going to be greater than zero. This might get flagged by something like CodeQL later.

If you're going to subtract and expect subtraction to go below zero, then this needs to be an `INTN`.

Reply:
0 should be very certain regardless of the user input because it's coming from the GetVariable(..) api. This would mean that the platform is wildly out of spec and I don't think I can handle that.